services/storage/tests/unit/test_handlers_paths.py

```diff
@@ -0,0 +1,363 @@
+        client,
+        location_id,
+        user_id,
+        file_filter=None,
+        expected_paths=[],
+    )
+
+
+@pytest.mark.parametrize(
+    "project_params",
+    [
+        ProjectWithFilesParams(
+            num_nodes=10,
+            allowed_file_sizes=(TypeAdapter(ByteSize).validate_python("1b"),),
+            workspace_files_count=10,
+        )
+    ],
+    ids=str,
+)
+async def test_list_paths_pagination(
+    initialized_app: FastAPI,
+    client: httpx.AsyncClient,
+    location_id: LocationID,
+    user_id: UserID,
+    with_random_project_with_files: tuple[
+        dict[str, Any],
+        dict[NodeID, dict[SimcoreS3FileID, FileIDDict]],
+    ],
+    faker: Faker,
+):
+    project, list_of_files = with_random_project_with_files
+    num_nodes = len(list(project["workbench"]))
+
+    # ls the nodes (DB-based)
+    file_filter = Path(project["uuid"])
+    expected_paths = sorted(
+        ((file_filter / node_key, False) for node_key in project["workbench"]),
+        key=lambda x: x[0],
+    )
+    await _assert_list_paths(
+        initialized_app,
+        client,
+        location_id,
+        user_id,
+        file_filter=file_filter,
+        expected_paths=expected_paths,
+        limit=int(num_nodes / 2 + 0.5),
+    )
+
+    # ls in the workspace (S3-based)
+    # ls in the workspace
+    selected_node_id = NodeID(random.choice(list(project["workbench"])))  # noqa: S311
+    selected_node_s3_keys = [
+        Path(s3_object_id) for s3_object_id in list_of_files[selected_node_id]
+    ]
+    workspace_file_filter = file_filter / f"{selected_node_id}" / "workspace"
+    expected_paths = _filter_and_group_paths_one_level_deeper(
+        selected_node_s3_keys, workspace_file_filter
+    )
+    await _assert_list_paths(
+        initialized_app,
+        client,
+        location_id,
+        user_id,
+        file_filter=workspace_file_filter,
+        expected_paths=expected_paths,
+        limit=1,
+        check_total=False,
+    )
+    # ls in until we get to some files
+    while selected_subfolders := [p for p in expected_paths if p[1] is False]:
+        selected_path_filter = random.choice(selected_subfolders)  # noqa: S311
+        expected_paths = _filter_and_group_paths_one_level_deeper(
+            selected_node_s3_keys, selected_path_filter[0]
+        )
+        await _assert_list_paths(
+            initialized_app,
+            client,
+            location_id,
+            user_id,
+            file_filter=selected_path_filter[0],
+            expected_paths=expected_paths,
+            check_total=False,
+        )
+
+
+@pytest.mark.parametrize(
+    "project_params, num_projects",
+    [
+        (
+            ProjectWithFilesParams(
+                num_nodes=3,
+                allowed_file_sizes=(TypeAdapter(ByteSize).validate_python("1b"),),
+                workspace_files_count=10,
+            ),
+            3,
+        )
+    ],
+    ids=str,
+)
+async def test_list_paths(
```

Comment:
i understand this is an entire workflow
MICRO: would it make sense to split in smaller tests (later?)

Reply:
not sure. i would have to look into it.